cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -827,30 +840,33 @@ shuffle_and_organize_output(
       auto mem_frugal_threshold = static_cast<size_t>(
         static_cast<double>(total_global_mem / element_size) * mem_frugal_ratio);
 
-      auto d_tx_value_counts = cugraph::groupby_and_count(
-        labels->begin(),
-        labels->end(),
-        [output_label = std::get<0>(*label_to_output_comm_rank),
-         output_rank  = std::get<1>(*label_to_output_comm_rank)] __device__(auto val) {
-          auto pos =
-            thrust::lower_bound(thrust::seq, output_label.begin(), output_label.end(), val);
-          return output_rank[thrust::distance(output_label.begin(), pos)];
-        },
-        comm_size,
-        mem_frugal_threshold,
-        handle.get_stream());
-
-      std::vector<size_t> h_tx_value_counts(d_tx_value_counts.size());
-      raft::update_host(h_tx_value_counts.data(),
-                        d_tx_value_counts.data(),
-                        d_tx_value_counts.size(),
-                        handle.get_stream());
-      handle.sync_stream();
-
       if (weights) {
         if (edge_ids) {
           if (edge_types) {
             if (hops) {
+              auto d_tx_value_counts = cugraph::groupby_and_count(
+                labels->begin(),
+                labels->end(),
+                thrust::make_zip_iterator(majors.begin(),
+                                          minors.begin(),
+                                          weights->begin(),
+                                          edge_ids->begin(),
+                                          edge_types->begin(),
+                                          hops->begin()),
+                shuffle_to_output_comm_rank_t<label_t>{std::get<0>(*label_to_output_comm_rank),
+                                                       std::get<1>(*label_to_output_comm_rank)},
+                comm_size,
+                mem_frugal_threshold,
+                handle.get_stream());
+
+              std::vector<size_t> h_tx_value_counts(d_tx_value_counts.size());
+              raft::update_host(h_tx_value_counts.data(),
+                                d_tx_value_counts.data(),
+                                d_tx_value_counts.size(),
+                                handle.get_stream());
+
+              handle.sync_stream();
```

Comment:
We can call `cugraph::groupby_and_count` only once before exponential if/else checks if we use an index array and then use ` `thrust::gather`  or thrust::permutation_iterator`  to rearrange 

```
majors
minors
weights
edge_ids
edge_types
hops
```
using  a zip iterator.

Reply:
Or isn't it possible to make the body of each if-else cases as a template function. If I am not mistaken, the only differences between the cases is the iterators we are zipping using `thrust::make_zip_iterator(...)`

Reply:
the only differences between the cases is the iterators we are zipping using thrust::make_zip_iterator(...) - that's correct. 
It can be merged as it and can be changed at later point as Chuck might be away.